### ISSUE TITLE
websocket writes -- no feedback when the write is complete

### ISSUE THREAD

**Issue by netikras** [kubernetes-client/java#1310]
Whenever a WRITE is issued to the attached container (e.g. file upload, long command, etc.), the data-to-be-written is added to the write queue, which is drained asynchronously by OKHttp3. And there is no way to get feedback when the write is complete. If I'm writing an integration with k8s, I'd like to know when my file has been uploaded, so that I could go to the next step and use it in the container. OFC I could do Thread.sleep(120_000L), but it is not exactly the right way to go IMO...

Please either make the writes synchronous, so the user can handle the concurrency flows himself, or add an ability to get feedback when the write is complete (a Future<?> would be nice). Without this feedback, any commands sent to the pod and any file uploads are subject for race condition without any way to apply synchronization techniques.

**Comment by netikras** [comment]
Related: https://github.com/kubernetes-client/java/issues/1127

**Comment by brendandburns** [comment]
Yep, this is fair. I assume that you are using the `Copy` helper class?

**Comment by netikras** [comment]
@brendanburns I was using a custom Copy implementation, based on Exec and `dd if=/dev/stdin` (just like `Copy` does), because 
- I cannot guarantee all the pods will have base64 installed
- I need to upload files from the working memory - NOT from the filesystem, and `Copy` only accepts a `Path` as a source (Why not InputStream though?)
- If the pod is unable to accept data with `dd` - I can swap the acceptance mechanism with any other command, reading from stdin (e.g. tee)
- lack of _append_ functionality in `Copy`


And just now I have tried the `Copy` utility and I can confirm that it is also uploading data completely asynchronously.

P.S. I have found kind of a workaround for that - caching the command's onClose event and completing the future then. However, this approach refrains me from reusing the connection/stream multiple times.
Since I have a sort-of working workaround now [the `Copy` is still not a subject for], the 16MiB problem [in another ticket I've raised today] is a lot more severe than this one.

**Comment by netikras** [comment]
BTW, what's with the base64 in the `Copy` util? Why compress the data and then explode it with b64? WebSockets' should be perfectly capable to deal with binary streams AFAIK?

**Comment by brendandburns** [comment]
In general, I agree with your feedback that we need to have a `Future` to represent the completion of an Exec. You should be able to wait on the `Process` to complete though which should indicate that it's done.

Regarding base64, I agree it is probably unnecessary, it was used for single file cases since it's easier than tarballing a single file, but for multiple files it's probably not needed.

**Comment by netikras** [comment]
Yepp, waiting for the process to exit is actually what I'm doing. Too bad the utilities ( `Copy`, `PortForward`, etc.) do not have a mechanism to monitor this event.

This is the wrapper for Kube's Exec I've come up with. I didn't find much use for the `Process` - it's only the `Process` events and streams that actually matter. So I've hidden the Process as an implementation detail and only exposed what really matters. Especially when it's a dehydrated Process and violates the Liskov's principle - not implemented methods throw `java.lang.UnsupportedOperationException`, which is a ticking bomb in the application.

```
    @Override
    public Future<Integer> exec(String namespace, String podName, Consumer<IOTrio> onOpen, BiConsumer<Integer, IOTrio> onClosed, BiConsumer<Throwable, IOTrio> onError, Long timeoutMs, boolean tty, String... command) throws IOException {
        CompletableFuture<Integer> future = new CompletableFuture<>();
        io.kubernetes.client.Exec exec = new io.kubernetes.client.Exec(getClient(namespace));
        IOTrio io = new IOTrio();
        String cmdStr = Arrays.toString(command);
        try {
            Process process = exec.newExecutionBuilder(resolveNamespace(namespace), podName, command)
                    .setStdin(true)
                    .setStdout(true)
                    .setStderr(true)
                    .setTty(tty)
                    .execute();
            io.onClose((code, timeout) -> {
                process.destroy();
                try {
                    process.waitFor(timeout, TimeUnit.MILLISECONDS);
                } catch (Exception e) {
                }
                if (process.isAlive())
                    process.destroyForcibly();
            });
            io.setStdin(process.getOutputStream());
            io.setStdout(process.getInputStream());
            io.setStderr(process.getErrorStream());
            process.onExit().thenAccept(p -> {
                logger.debug("process.onExit(): {}", cmdStr);
                future.complete(p.exitValue());
                onClosed.accept(p.exitValue(), io);
            });
            Thread processWaitingThread = new Thread(() -> {
                try {
                    logger.debug("Waiting for process to close for {} ms: {}", timeoutMs, cmdStr);
                    if (timeoutMs != null) {
                        boolean exited = process.waitFor(timeoutMs, TimeUnit.MILLISECONDS);
                        logger.debug("Process closed={}: {}", exited, cmdStr);
                        if (!exited && process.isAlive()) {
                            process.destroyForcibly();
                            future.completeExceptionally(new TimeoutException());
                        }
                    } else {
                        process.waitFor();
                    }
                } catch (Exception e) {
                    onError.accept(e, io);
                }
            });
            processWaitingThread.setName("Process-Waiting-Thread-" + command[0] + "-" + podName);
            processWaitingThread.start();
            onOpen.accept(io);
            return future;
        } catch (ApiException e) {
            onError.accept(e, io);
            future.completeExceptionally(e);
        }
        return future;
    }
```


As for base64 - I'm still not following. I'm uploading non-textual files in their binary format (using `dd`) w/o any issues - I can't think of a case where b64 would give any advantage, be it a single file or multiple files. If anything - b64 increases the payload size by ~30%.

**Comment by brendandburns** [comment]
That's fair, I think we probably thought some encoding was better than raw binary, but it's not clear that's a good rationale.

**Comment by brendandburns** [comment]
I think we'd be happy to accept a PR adding that `Future<int> exec(...)` method to the `Exec` class.
